### PR TITLE
Actions Artifacts v4 backend

### LINKED ISSUE #28853: Support actions/upload-artifact@v4, actions/download-artifact@v4
### Feature Description

The v4 has a different protocol than before.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by ChristopherHX**
Fixes #28853 

Needs both https://gitea.com/gitea/act_runner/pulls/473 and https://gitea.com/gitea/act_runner/pulls/471 on the runner side and patched `actions/upload-artifact@v4` / `actions/download-artifact@v4`, like `christopherhx/gitea-upload-artifact@v4` and `christopherhx/gitea-download-artifact@v4`, to not return errors due to GHES not beeing supported yet.

---
Known issues

- official action v4 doesn't allow replace, this is not checked
  - to check this, we need to know if the artifact has been created in the current run_attempt of the job
- added fixture uses some invalid task token data, which are not used by the test and required by constraints
  - I'm not shure how to generate these fields for a fixture

Open Questions

- Can we Append Blockstorage?
  - merging chunks like now is inefficient
  - use the url of the backend to upload via a signed url?
  - GitHub returns the azure Blockstorage url for direct upload without any proxy like used in this PR

Workflow used for testing purposes
```yaml
on:
  push:
jobs:
  _5:
    runs-on: self-hosted
    steps: 
    - run: env
    - run: |
        github:
        ${{ tojson(github) }}
        inputs:
        ${{ tojson(inputs) }}
        matrix:
        ${{ tojson(matrix) }}
        needs:
        ${{ tojson(needs) }}
        strategy:
        ${{ tojson(strategy) }}            
      shell: cp {0} context.txt
    - uses: christopherhx/gitea-upload-artifact@v4
      with:
        name: test
        path: context.txt

    - uses: christopherhx/gitea-download-artifact@v4
      with:
        name: test
        path: out
    - run: cat out/context.txt
 ```

preview

### CODE REVIEW

File: routers/web/repo/actions/view.go
Comment (ChristopherHX) [inline routers/web/repo/actions/view.go]: Removing this just leads to a double zipped download, otherwise can be removed

File: routers/api/actions/artifacts_chunks.go
Comment (wxiaoguang) [inline routers/api/actions/artifacts_chunks.go]: It looks like duplicate code.

Not a blocker, but if there is a chance to reduce duplication, I think it's worth to try.

File: routers/api/actions/artifactsv4.go
Comment (wxiaoguang) [inline routers/api/actions/artifactsv4.go]: A nit, all other places use `http.StatusXxx` constants.

File: routers/api/actions/artifactsv4.go
Comment (wxiaoguang) [inline routers/api/actions/artifactsv4.go]: IIRC `ctx.Error` also prints error log, so all `log.Error` could be simplified?

**Comment by kgdev** [comment]
Does it support ServeDirect when downloading an artifact?

**Comment by lunny** [comment]
> Does it support ServeDirect when downloading an artifact?

It's another issue. Please fire another issue.

File: routers/api/actions/artifactsv4.go
Comment (ChristopherHX) [reply routers/api/actions/artifactsv4.go]: I think you mean ApiContext, here we have context.Base that doesn't seem to write anything to log.

Create a new ArtifactContext struct which uses ApiContext instead? the current struct is shared with the old artifacts code, so I would rather not change the existing ArtifactContext that much in this PR. I have no problem in doing this in another PR for the old backend.

File: routers/api/actions/artifactsv4.go
Comment (ChristopherHX) [reply routers/api/actions/artifactsv4.go]: fixed

File: routers/api/actions/artifacts_chunks.go
Comment (ChristopherHX) [reply routers/api/actions/artifacts_chunks.go]: you are correct about this, I have had to get rid of the md5 and http range logic of the existing code path

disabling these features using some ifs is also working

File: routers/api/actions/artifactsv4.go
Comment (wxiaoguang) [reply routers/api/actions/artifactsv4.go]: Hmm ... maybe it's better to leave it to the future .....

File: routers/api/actions/artifactsv4.go
Comment (wxiaoguang) [inline routers/api/actions/artifactsv4.go]: Is `setting.SecretKey` good to be used here? It has a default value and it is the same for most instances. (ps: SecretKey was never seriously used)

File: routers/web/repo/actions/view.go
Comment (wxiaoguang) [reply routers/web/repo/actions/view.go]: How other other platforms handle this case?

There are similar problems for "zip in zip": "gz in zip", "xz in zip", "bz in zip", etc.

Without deep thought, the brief feeling: for consistent behavior: I think it doesn't need to "bypass" the zip-in-zip, just provide the files in a consistent way: always zip the file. Or, add a parameter to let end users decide whether they need to zip?

Comment (ChristopherHX) [reply routers/web/repo/actions/view.go]: > How other other platforms handle this case?

I'm trying to explain some differences between v3 and v4, you might not know

On GitHub
- we can download the single zip artifact uploaded via v4
- we can download the zipped individual files via v1-v3

Notes|v1-v3|v4
---|---|---
within workflow|uploads individual files|uploads exactly one zip containing all files
within workflow|downloads individual files in workflows|downloads exactly one zip per artifact
after workflow|download zip of uploaded files (that's what the code here is doing)|download the single zip uploaded from blob storage


> There are similar problems for "zip in zip": "gz in zip", "xz in zip", "bz in zip", etc.

These are not handled at all. If v3 uploads a single zip, a double zipped download is expected on my end.

**BTW** I have not tried to upload a xz to the blob storage of the GitHub Action V4 artifact (Requires modding of the action)

`ContentEncoding` is currently never `application/zip` in the v3 backend, used for detecting gz storage of a single file

I agree that this is suboptimal
- I could change the code to extract the zip and store induvidual files in v4
  - Then zip on demand on download within workflows
  - would break direct download from blob storage using sign url
  - would make v3 and v4 compatible on gitea while they are not in GitHub
- We could split out the artifact storage so we clearly know wether it is a v1-v3 artifact, which needs to be zipped, or v4 which is stored as zio

> Or, add a parameter to let end users decide whether they need to zip?

I don't think this would help here

Comment (wxiaoguang) [reply routers/web/repo/actions/view.go]: Thank you very much, that's really a complicated background ....

Maybe some information could be added as comment.

File: routers/api/actions/artifactsv4.go
Comment (ChristopherHX) [reply routers/api/actions/artifactsv4.go]: `setting.SecretKey` is also used for signing jwt tokens in packages and artifacts (I copies the approuch from packages)

Could you suggest an alternative key to use?
What comes into my mind would be
- using fields making up the validation of Action Task tokens, like the hash,salt or last chars of the token

We cannot use the other common auth methods, due to beeing called without any Auth Header by the client library.

Comment (wxiaoguang) [reply routers/api/actions/artifactsv4.go]: Hmm, no idea from my side at the moment. Maybe TOC could help.

File: routers/web/repo/actions/view.go
Comment (ChristopherHX) [reply routers/web/repo/actions/view.go]: I have added a small comment about the reasons for this condition

https://github.com/go-gitea/gitea/pull/28965/commits/6f2e5505932c3028aa3116998e2dca1353c79e86

I'm open for discussions, since I have never worked on a public facing api of Gitea.

File: routers/api/actions/artifactsv4.go
Comment (fuxiaohei) [inline routers/api/actions/artifactsv4.go]: The deletion removes record but not storage file. It provides logic to handle in https://github.com/go-gitea/gitea/pull/27172. I will make it merged soon.

File: routers/api/actions/artifactsv4.go
Comment (ChristopherHX) [reply routers/api/actions/artifactsv4.go]: I see you have added `setting.GetGeneralTokenSigningSecret()`, assume it is better than `setting.SecretKey` for the time being

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Review by fuxiaohei — APPROVED** [review]

File: routers/api/actions/artifactsv4.go
Comment (ChristopherHX) [reply routers/api/actions/artifactsv4.go]: I assume this has been resolved with <https://github.com/go-gitea/gitea/pull/28965/commits/ff30004285ad35190768d1ce787df03a5d59e578>

**Comment by wxiaoguang** [comment]
> I have noticed in my previous PR that the `---` is not respected to mark content that should not be part of the commit message during PR merge. So this additional content has been manually removed from my side.

I have called for "please edit the commit message manually before merge" in the member's discord channel.

But I guess such initiative is seldom followed. (At least, when I merge, I always edit).

**Pull request merged**
